File: 2022/day06/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Basename;
use File::Spec;

sub find_marker {
    my ($data, $window_size) = @_;
    my $len = length($data);

    for my $i ($window_size .. $len) {
        my $window = substr($data, $i - $window_size, $window_size);
        my %seen;
        $seen{$_}++ for split //, $window;
        if (scalar(keys %seen) == $window_size) {
            return $i;
        }
    }
    return -1;
}

sub part1 {
    my ($data) = @_;
    return find_marker($data, 4);
}

sub part2 {
    my ($data) = @_;
    return find_marker($data, 14);
}

sub main {
    my $script_dir = dirname(__FILE__);
    my $input_file = File::Spec->catfile($script_dir, '..', 'input.txt');

    open my $fh, '<', $input_file or die "Cannot open $input_file: $!";
    my $data = do { local $/; <$fh> };
    close $fh;

    $data =~ s/\s+$//;  # Strip trailing whitespace

    print "Part 1: ", part1($data), "\n";
    print "Part 2: ", part2($data), "\n";
}

main();

File: 2022/day06/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <unordered_set>

int find_marker(const std::string& data, int window_size) {
    for (size_t i = window_size; i <= data.length(); ++i) {
        std::unordered_set<char> window(data.begin() + i - window_size, data.begin() + i);
        if (window.size() == static_cast<size_t>(window_size)) {
            return static_cast<int>(i);
        }
    }
    return -1;
}

int part1(const std::string& data) {
    return find_marker(data, 4);
}

int part2(const std::string& data) {
    return find_marker(data, 14);
}

int main() {
    std::ifstream file("../input.txt");
    if (!file.is_open()) {
        std::cerr << "Error: Could not open input file" << std::endl;
        return 1;
    }

    std::string data;
    std::getline(file, data);
    file.close();

    std::cout << "Part 1: " << part1(data) << std::endl;
    std::cout << "Part 2: " << part2(data) << std::endl;

    return 0;
}
